### PR TITLE
[infra] Refactor Vulkan runtime into true Common Runtime

### PR DESCRIPTION
**Pull request by bobcao3**
After this PR, the previous Vulkan runtime is renamed to the "common runtime" for unified device api, under namespace `taichi::lang::gfx` and folder `taichi/runtime/gfx`

In addition, related AOT modules code are moved into the runtime as well, as they do not actually have a Vulkan dependency. This also prepares for moving OpenGL backend to the common runtime. (Currently DX11 is already using the common runtime)

Also this PR fixes a small implementation issue in OpenGL where `rw_buffer` should be used in resource binder for SSBO instead of `buffer` (which is for UBOs). In addition, this PR fixes the memory leak issue in DX11. Although the DX11 device API is still quite lacking in features, it is now as usable as a base Vulkan 1.1 implementation.

This PR also changed linker flags a bit for Windows, now LTO is only enabled for release builds (debug builds will not use LTO)

### CODE REVIEW

File: taichi/gui/gui.h [hunk 1]
```diff
@@ -888,7 +888,7 @@ class TI_DLL_EXPORT GUI : public GUIBase {
       redraw();
       process_event();
       if (frame_id % 10 == 0)
-        set_title(fmt::format("{} ({:.2f} FPS)", window_name, real_fps));
+        set_title(fmt::format("{} ({:.2f} FPS)", "", real_fps));
```
Comment (ailzhang) [inline taichi/gui/gui.h:891]: Why's this change?

File: taichi/aot/module_loader.cpp [hunk 2]
```diff
@@ -32,19 +32,22 @@ Kernel *KernelTemplate::get_kernel(
...
-#else
-    TI_NOT_IMPLEMENTED
+    return gfx::make_aot_module(mod_params, arch);
+#endif
+  } else if (arch == Arch::dx11) {
+#ifdef TI_WITH_DX11
+    return gfx::make_aot_module(mod_params, arch);
+#endif
+  } else if (arch == Arch::opengl) {
+#ifdef TI_WITH_OPENGL
```
Comment (ailzhang) [inline taichi/aot/module_loader.cpp:42]: opengl is not yet supported? Shall we leave that out of this PR?

**Review by ailzhang — COMMENTED** [review]
Overall looks good to me! Leaving this to @k-ye to take a second look before merging.

File: taichi/gui/gui.h (hunk 1 above)
Comment (bobcao3) [reply taichi/gui/gui.h:891]: Accidentally included. The text on my Windows 10 is all garbled, tried to fix it but no luck yet

File: taichi/runtime/gfx/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,16 @@
+# ./taichi/runtime/gfx/CMakeLists.txt
+
+add_library(gfx_runtime)
```
Comment (k-ye) [inline taichi/runtime/gfx/CMakeLists.txt:3]: neat!

File: taichi/runtime/gfx/snode_tree_manager.h [hunk 4]
```diff
@@ -8,9 +8,9 @@
...
 namespace taichi {
 namespace lang {
-namespace vulkan {
+namespace gfx {
 
-class VkRuntime;
+class GfxRuntime;
 
 /**
  * @brief Manages the SNodeTrees for the Vulkan backend.
```
Comment (k-ye) [inline taichi/runtime/gfx/snode_tree_manager.h:16]:
```suggestion
 * @brief Manages the SNodeTrees for the underlying backend.
```

**Review by k-ye — APPROVED** [review]
LGTM!

What's the plan for OpenGL? Is it to derive a new runtime from `gfx::GfxRuntime`?

File: taichi/runtime/gfx/CMakeLists.txt (hunk 3 above)
Comment (bobcao3) [reply taichi/runtime/gfx/CMakeLists.txt:3]: Yes, that's the plan

**Review by qiao-bo — APPROVED** [review]
LGTM

**Pull request merged**